File: pset6/readability/readability.py
import re


def main():
    # Prompt user for text input
    text = input("Text: ")

    # text length
    length = len(text)
    letters = 0
    words = 1
    sentences = 0

    # Check for letters in text
    for i in range(length):
        # Check if char is an alphabet character
        if text[i].isalpha():
            letters += 1

    # Check for words in text
    for i in range(length):
        # Check if char is space
        if text[i].isspace():
            words += 1

    regex = re.compile('[.!?]')
    # Check for sentences in text
    for i in range(length):
        # Check if char is is the regex
        if (regex.search(text[i]) != None):
            sentences += 1

    # Compute the grade using Coleman-Liau algo
    L = 100 * (letters / words)
    S = 100 * (sentences / words)
    grade = round(0.0588 * L - 0.296 * S - 15.8)

    # Check the grade and print the appropriate message
    if grade < 1:
        print("Before Grade 1")
    
    elif grade > 16:
        print("Grade 16+")
    
    else:
        print(f"Grade {grade}")
    

main()

File: pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>
#include <string.h>

int main(void)
{
    string text = get_string("Text: ");

    int length = strlen(text);

    float letters = 0;
    for (int i = 0, n = length; i < n; i++)
    {
        if ((text[i] >= 97 && text[i] <= 122) ||
            (text[i] >= 65 && text[i] <= 90))
        {
            letters++;
        }
    }

    float words = 1;
    for (int i = 0, n = length; i < n; i++)
    {
        if (text[i] == 32)
        {
            words++;
        }
    }

    float sentences = 0;
    for (int i = 0, n = length; i < n; i++)
    {
        if (text[i] == 46 || text[i] == 33 || text[i] == 63)
        {
            sentences++;
        }
    }

    // Calculate average number of letters & sentences per 100 words
    float L = 100 * (letters / words);
    float S = 100 * (sentences / words);

    int grade = round(0.0588 * L - 0.296 * S - 15.8);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}